compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlCodeToTransaction.cpp

```diff
@@ -63,20 +63,37 @@ LogicalResult convertOp(AMDAIE::NpuWriteBdOp op, TransactionBuilder &builder) {
   return success();
 }
 
+LogicalResult convertOp(AMDAIE::DMAStartOp op, TransactionBuilder &builder) {
+  if (failed(builder.appendDmaStartOp(op))) {
+    return failure();
+  }
+  return success();
+}
+
 LogicalResult controlCodeToTransaction(IRRewriter &rewriter,
                                        AMDAIE::ControlCodeOp controlCodeOp,
                                        TransactionBuilder &builder) {
   SmallVector<Operation *> toBeErased;
-  WalkResult res = controlCodeOp->walk([&](Operation *op) {
+  DenseSet<AMDAIE::LockOp> lockOps;
+  WalkResult res = controlCodeOp->walk([&](AMDAIE::UseLockOp op) {
+    auto lockOp = op.getLock().getDefiningOp<AMDAIE::LockOp>();
+    if (lockOps.count(lockOp) == 0) {
+      if (failed(builder.appendLockOp(lockOp))) return WalkResult::interrupt();
+      lockOps.insert(lockOp);
+    }
```

Comment:
```suggestion
    if (lockOps.contains(lockOp)) return WalkResult::advance();
    if (failed(builder.appendLockOp(lockOp))) return WalkResult::interrupt();
    lockOps.insert(lockOp);
```
You can avoid the nested condition here.